File: src/main/java/com/duxtinto/intellij/plugin/github/codereviews/ide/acl/services/GithubRepositoryFinderImpl.java
package com.duxtinto.intellij.plugin.github.codereviews.ide.acl.services;

import com.duxtinto.intellij.plugin.github.codereviews.domain.repositories.GithubRepositoryEntity;
import com.duxtinto.intellij.plugin.github.codereviews.domain.repositories.RepositoriesDomainContract;
import com.duxtinto.intellij.plugin.github.codereviews.ide.acl.entities.GitRepositoryExt;
import com.intellij.openapi.util.Pair;
import git4idea.repo.GitRemote;
import org.jetbrains.plugins.github.api.GithubFullPath;
import org.jetbrains.plugins.github.util.GithubUrlUtil;
import org.jetbrains.plugins.github.util.GithubUtil;

import javax.annotation.Nullable;
import javax.inject.Inject;

public class GithubRepositoryFinderImpl implements RepositoriesDomainContract.GithubRepositoryFinder {
    @Inject
    public GithubRepositoryFinderImpl() {
    }

    @Override
    @Nullable
    public GithubRepositoryEntity find(GitRepositoryExt repository) {
        Pair<GitRemote, String> remote = GithubUtil.findGithubRemote(repository.delegate());
        if (remote == null) {
            return null;
        }

        GithubFullPath path = GithubUrlUtil.getUserAndRepositoryFromRemoteUrl(remote.getSecond());
        if (path == null) {
            return null;
        }

        return new GithubRepositoryEntity(path.getRepository(), path.getUser());
    }
}

File: src/main/kotlin/com/duxtinto/intellij/plugin/github/codereviews/ide/acl/services/GithubRepositoryFinderImpl.kt
package com.duxtinto.intellij.plugin.github.codereviews.ide.acl.services

import com.duxtinto.intellij.plugin.github.codereviews.domain.repositories.GithubRepositoryEntity
import com.duxtinto.intellij.plugin.github.codereviews.domain.repositories.RepositoriesDomainContract
import com.duxtinto.intellij.plugin.github.codereviews.ide.acl.entities.GitRepositoryExt
import org.jetbrains.plugins.github.util.GithubUrlUtil
import org.jetbrains.plugins.github.util.GithubUtil
import javax.inject.Inject

class GithubRepositoryFinderImpl @Inject
constructor() : RepositoriesDomainContract.GithubRepositoryFinder {

    override fun find(gitRepository: GitRepositoryExt): GithubRepositoryEntity? {
        val remote = GithubUtil.findGithubRemote(gitRepository.delegate()) ?: return null
        val path = GithubUrlUtil.getUserAndRepositoryFromRemoteUrl(remote.getSecond()) ?: return null

        return GithubRepositoryEntity(path.repository, path.user)
    }
}
